quill-jdbc/src/main/scala/io/getquill/context/jdbc/JdbcEncoders.scala

```diff
@@ -26,56 +30,44 @@ trait JdbcEncoders {
         }
         row.single(index, value, raw)
       }
-    }
+    })
 
   implicit def traversableEncoder[T](implicit enc: Encoder[T]): Encoder[Traversable[T]] =
     new Encoder[Traversable[T]] {
       override def apply(index: Int, values: Traversable[T], row: BindedStatementBuilder[PreparedStatement]) =
         row.coll(index, values, enc)
     }
 
-  private[this] val nullEncoder = encoder[Int](_.setNull)
+  private[this] val nullEncoder = encoder[Int](_.setNull, Types.INTEGER)
 
   implicit def optionEncoder[T](implicit d: Encoder[T]): Encoder[Option[T]] =
     new Encoder[Option[T]] {
       override def apply(idx: Int, value: Option[T], row: BindedStatementBuilder[PreparedStatement]) =
         value match {
           case Some(value) => d(idx, value, row)
-          case None =>
-            import Types._
-            val sqlType =
-              d match {
-                case `stringEncoder`     => VARCHAR
-                case `bigDecimalEncoder` => NUMERIC
-                case `booleanEncoder`    => BOOLEAN
-                case `byteEncoder`       => TINYINT
-                case `shortEncoder`      => SMALLINT
-                case `intEncoder`        => INTEGER
-                case `longEncoder`       => BIGINT
-                case `floatEncoder`      => REAL
-                case `doubleEncoder`     => DOUBLE
-                case `byteArrayEncoder`  => VARBINARY
-                case `dateEncoder`       => TIMESTAMP
-              }
-            nullEncoder(idx, sqlType, row)
+          case None => d match {
+            case JdbcEncoder(sqlType) => nullEncoder(idx, sqlType, row)
+            case _                    => throw new NotImplementedError("")
```

Comment:
Please use `fail` from `MessageUtil`. Also, could you provide a descriptive error message?